### DISCUSSION TITLE
Can Cypress customize `Host` header by force?

### DISCUSSION THREAD

**Discussion by gyh2016** [cypress-io/cypress#24819] [Questions and Help]
I have read the explanation about the two sources of the `Host` header(actually only one from req.url) in https://github.com/cypress-io/cypress/issues/14920.

But in my test case, I really need to modify the `Host` header to a different host. For example:

```
cy.intercept({ url: "some_path" }, (req) => {
  req.headers["host"] = "bbb.com";
  req.url = "aaa.com";
  req.reply();

  // or use redirect
  // req.redirect("aaa.com", 302);
}).as("some_alias");
```

When visit aaa.com, it only works when the `Host` header is equal to "bbb.com".

`curl` can help me do that. It supports set `Host` header by `-H` parameter such as `curl "aaa.com" -H "host:bbb.com"`.

Does Cypress support to customize `Host` header by force like this?

**Comment by portyu9** [comment]
Hi @gyh2016 

For **browser traffic**, I would not rely on `cy.intercept()` to send a request to `aaa.com` while forcing its HTTP `Host` to `bbb.com`.

`Host` is special: it identifies the destination virtual host and is controlled by the browser/network stack. This is different from modifying normal headers such as `Authorization` or a custom `X-*` header.

Also, `req.redirect()` does not change how Cypress routes the original upstream request — it creates a redirect response that the browser then follows.

If this is an **API-only request**, `cy.request()` is a better layer because it is made from Cypress's Node-side HTTP client rather than from the application browser:

```js
cy.request({
  url: 'https://aaa.com/some_path',
  headers: {
    Host: 'bbb.com',
  },
})
```

For actual browser/E2E traffic, the robust solution is to make the URL use the virtual host the server expects, for example by configuring test DNS, `/etc/hosts`, or a reverse proxy so that `bbb.com` resolves/routes to the `aaa.com` backend.

That keeps the URL, TLS/SNI, and HTTP host routing consistent instead of trying to override a transport-level header after the fact.

Please mark this answer as accepted if it helped, thank you.
